### PR TITLE
Add HTTP PUT support for OSCAL System Security Plans

### PR DESCRIPTION
**Pull request by pkothare**
## Motivation
The OSCAL REST API should allow for wholesale replacement of a system security plan (SSP). Clients will want to submit an SSP that may have updates across multiple nested objects. Creating a diff and subsequently an exhaustive list of `PATCH` operations requires significant overhead in maintaining state. This may be unwarranted for simple clients that want to implement high-level operations against an instance of the [OSCAL SSP model](https://pages.nist.gov/OSCAL/concepts/layer/implementation/ssp/).

## Specification
To accommodate the scenario described above, this update adds support the for HTTP `PUT` method under the path:
```
/system-security-plans/{sspId}
```
where `sspId` is the [System Security Plan Universally Unique Identifier](https://pages.nist.gov/OSCAL/reference/latest/system-security-plan/json-reference/#/system-security-plan/uuid). Furthermore, a request body containing the **entire** JSON SSP model is also required to fulfill this request. ~~Lastly, the `UUID` value specified in the body of the request under the JSON path `$.system-security-plan.uuid` is ignored under all circumstances, because the `sspId` specified in URI path takes precedence when aligning with HTTP/1.1 semantics. Since a `UUID` is required to pass schema validation, any `UUID` specified must be accepted, but not processed.~~ 

### HTTP/1.1 Semantics Alignment
Based on quoted sections from the [PUT method definition](https://datatracker.ietf.org/doc/html/rfc7231#section-4.3.4), this update also includes sematic alignment for the `PUT` request via HTTP status codes.

~~In the case when the origin (server) cannot find the SSP using the `sspId`, the origin must create a new SSP resource and return a `201` response code. This method is not a replacement for the `POST` method since it is idempotent. To maintain idempotency, the `sspId` specified by the client should be honored by the origin. The method by which an origin handles conflicts of UUIDs is out of scope for this specification.~~

This scenario will not be supported.

> If the target resource does have a current representation and that representation is successfully modified in accordance with the state of the enclosed representation, then the origin server MUST send either a 200 (OK) or a 204 (No Content) response to indicate successful completion of the request.

The origin must return a `204` status code if it modified an existing SSP resource identified by the `sspId`. The `204` status code is preferred over the `200` status code since SSP JSON objects can be large, causing significant network bandwidth usage. Future work may include support for the `200` status code if versioning can be handled correctly.

- For this reason, a client must submit the **entire** SSP JSON object as part of the request body. `multipart` submissions are not allowed on this endpoint, and the origin must respond with a `400` status code, if the client mistakenly tries to submit such a request. 
- A `400` status code should also be returned in case the JSON that was submitted is malformed.
- If valid JSON was submitted, but the JSON object fails semantic validation (i.e. validation against the [OSCAL JSON SSP schema](https://pages.nist.gov/OSCAL/reference/latest/system-security-plan/json-outline/)) a `400` status code must be returned by the origin.

> When a PUT representation is inconsistent with the target resource, the origin server SHOULD either make them consistent, by transforming the representation or changing the resource configuration, or respond with an appropriate error message containing sufficient information to explain why the representation is unsuitable.  The 409 (Conflict) or 415 (Unsupported Media Type) status codes are suggested, with the latter being specific to constraints on Content-Type values.

- The origin should return a `415` status code if the `Content-Type` header is not `application/json`. Per [Section 3.1.1.5 of [RFC7231]](https://datatracker.ietf.org/doc/html/rfc7231#section-3.1.1.5), the origin may infer the media type if it is not specified to improve compatibility. However, if deemed incompatible the `415` status code must be returned.

## Future Considerations
- Adding support for `ETag` and `Last-Modified` based [Section 4.3.4 of [RFC7231]](https://www.rfc-editor.org/rfc/rfc7231.html#section-4.3.4). This should be carefully aligned with the [`metadata`](https://pages.nist.gov/OSCAL/reference/latest/system-security-plan/json-reference/#/system-security-plan/metadata) properties of an SSP.
- Adding support for the `409` error status code. This status code was intentionally left out since it requires a rigorous implementation of version tracking. Since version tracking itself is a consideration for future improvement, supporting the `409` status code cannot be accommodated.
- Adding support for the `422` error status code. While the `422` error status code is not part of RFC 7231, it is part of the draft [HTTP Semantics](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-semantics-19) standard. If approved, support for this error code, will enable clients to improve error handling.

## References
- [RFC 7231 - Hypertext Transfer Protocol (HTTP/1.1): Semantics and Content](https://datatracker.ietf.org/doc/html/rfc7231)
- [OSCAL SSP model](https://pages.nist.gov/OSCAL/concepts/layer/implementation/ssp/)
- [System Security Plan Model v1.0.1 JSON Format Reference](https://pages.nist.gov/OSCAL/reference/latest/system-security-plan/json-reference/#/system-security-plan/)
- [PUT - HTTP | MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PUT)
- https://github.com/EasyDynamics/easygrc/issues/23

### CODE REVIEW

**Review by laurelmay — CHANGES_REQUESTED** [review]
So overall, I like the thoroughness and the focus on complying with the HTTP specification and RESTful semantics; however, I do want to talk about a few things before merging.

## Existing PUT operations

We already have PUT operations for a few endpoints:
 - Component Definition Parties
 - Profile imported catalogs
 - Profile imported profiles
 - Profile parties
 - SSP components
 - SSP parties

Each of these, as part of the documentation returns a 404 if _either_ the top level resource (SSP, Component Definition, or profile) ID cannot be found **or** the resource's specific ID cannot be found. This implies that it is not permitted to specify the ID of a new resource as part of a PUT (so basically, can never return a `201 Created`). All of our PUT operations across the specification should have consistent behavior. Either `/system-security-plans/:sspId` should not create new resources and should instead 404 or we should update all existing PUTs to maintain a consistent API.

In general, I think we're in a sorta murky spot with this one _because_ we have to specify a UUID in the request body per the OSCAL schema.

## Handling Path ID != UUID

So I know this is really more of an implementation thing but I think it's such an important issue that we really want to address it as part of the specification (and we probably should have as part of PATCH as well). I am really tempted to suggest that any sort of a mismatch (or attempt to modify the UUID as part of the request body) should return a `400 Bad Request` (or some other 4xx error). We probably want to consider exactly how we want to handle move/rename semantics.

Again, this likely needs to be addressed for _all_ PUT operations.

## Moving Forward

So as far as how to move forward with this PR, I don't want to increase the scope of the ticket you're working on too much and say "we have to fix everything now"; however, I think we should at least talk about how to handle these, especially when it comes to consistency, correctness, and safety within the spec vs following typical HTTP and RESTful semantics.

**Comment by pkothare** [comment]
>Each of these, as part of the documentation returns a 404 if either the top level resource (SSP, Component Definition, or profile) ID cannot be found or the resource's specific ID cannot be found. This implies that it is not permitted to specify the ID of a new resource as part of a PUT (so basically, can never return a 201 Created)

If this was rationalized somewhere already, I don't want to go against the decisions of the team. Was there a decision log captured somewhere (I'll know where to look in the future)? If not, we can just use this PR as a means to formalize the decision. 

> So I know this is really more of an implementation thing but I think it's such an important issue that we really want to address it as part of the specification (and we probably should have as part of PATCH as well). I am really tempted to suggest that any sort of a mismatch (or attempt to modify the UUID as part of the request body) should return a 400 Bad Request (or some other 4xx error). We probably want to consider exactly how we want to handle move/rename semantics.

I agree, one more decision needs to be made here. To add some more color, RFC 7231 states:

> When a PUT representation is inconsistent with the target resource, the origin server SHOULD either make them consistent, by transforming the representation or changing the resource configuration, or respond with an appropriate error message containing sufficient information to explain why the representation is unsuitable. 

The design I opted for implied reconciling the PUT representation and the resource configuration, by ignoring the `uuid` in the request body. If we agree that the `sspId` in the path should match the `uuid` in the request body (and a `4xx` status code should be returned on mismatch), it still aligns with the HTTP spec. Not a huge difference one way or another for one operation, but as mentioned earlier, this could have a ripple effect on updating the spec of all other operations.

**Comment by rgauss** [comment]
> The design I opted for implied reconciling the PUT representation and the resource configuration, by ignoring the `uuid` in the request body. If we agree that the `sspId` in the path should match the `uuid` in the request body (and a `4xx` status code should be returned on mismatch), it still aligns with the HTTP spec. Not a huge difference one way or another for one operation, but as mentioned earlier, this could have a ripple effect on updating the spec of all other operations.

The o[verall description mentions the expected implementation behavior](https://github.com/EasyDynamics/oscal-rest/blob/develop/openapi.yaml#L46) around `PATCH` requests:
> The UUID of the root object must not change in the payload and implementations should throw an error if such a request is made.

So I think we'd want to suggest a similar behavior for `PUT`.

**Review by rgauss — COMMENTED** [review]
EasyDynamics/easygrc#23 is meant to address all currently supported 'root' OSCAL objects, i.e.:
- Catalogs
- Profiles
- Component Definitions
- System Security Plans

Should this PR be expanded to address those or will subsequent PRs follow?

**Comment by pkothare** [comment]
Added support for the `404` status code if the SSP cannot be found. @rgauss, I will submit additional PRs once I understand the design philosophy a little better.

**Review by laurelmay — COMMENTED** [review]
The `PUT /system-security-plans/:id` looks good! Sounds like we'll be adding this method to other routes in future PRs. Thanks for taking the time to address the feedback and discussion on this one!

**Review by laurelmay — APPROVED** [review]
Oops. My previous comment should have been an approval.

**Review by zclarkEDC — APPROVED** [review]

**Comment by laurelmay** [comment]
@pkothare you'll have to update this branch before we can merge

**Pull request merged**
